kernel/src/table_features/mod.rs

```diff
@@ -124,6 +150,424 @@ pub(crate) enum FeatureType {
+    Enabled(TableFeature),
+    /// Feature must NOT be supported
+    NotSupported(TableFeature),
+    /// Feature must NOT be enabled (may be supported but property must not activate it)
+    NotEnabled(TableFeature),
+    /// Custom validation logic
+    Custom(fn(&Protocol, &TableProperties) -> DeltaResult<()>),
+}
+
+/// Rich metadata about a table feature including version requirements, dependencies, and support status
+#[allow(dead_code)]
+#[derive(Clone)]
+pub(crate) struct FeatureInfo {
+    /// The feature's canonical name as it appears in the protocol
+    pub name: &'static str,
+    /// Minimum reader protocol version required for this feature
+    pub min_reader_version: i32,
+    /// Minimum writer protocol version required for this feature
+    pub min_writer_version: i32,
+    /// The type of feature (Writer, ReaderWriter, or Unknown)
+    pub feature_type: FeatureType,
+    /// Requirements this feature has (features + custom validations)
+    pub feature_requirements: &'static [FeatureRequirement],
+    /// Rust kernel's read support for this feature (may vary by Operation type)
+    pub read_support: KernelSupport,
+    /// Rust kernel's write support for this feature (may vary by Operation type)
+    pub write_support: KernelSupport,
+    /// How to check if this feature is enabled in a table
+    pub enablement_check: EnablementCheck,
+}
+
+// Static FeatureInfo instances for each table feature
+#[allow(dead_code)]
+static APPEND_ONLY_INFO: FeatureInfo = FeatureInfo {
+    name: "appendOnly",
+    min_reader_version: 1,
+    min_writer_version: 2,
+    feature_type: FeatureType::Writer,
+    feature_requirements: &[],
+    read_support: KernelSupport::NotSupported,
+    write_support: KernelSupport::Supported,
+    enablement_check: EnablementCheck::EnabledIf(|props| props.append_only == Some(true)),
+};
+
+#[allow(dead_code)]
+static INVARIANTS_INFO: FeatureInfo = FeatureInfo {
+    name: "invariants",
+    min_reader_version: 1,
+    min_writer_version: 2,
+    feature_type: FeatureType::Writer,
+    feature_requirements: &[],
+    read_support: KernelSupport::NotSupported,
+    write_support: KernelSupport::Supported,
+    enablement_check: EnablementCheck::AlwaysIfSupported,
+};
+
+#[allow(dead_code)]
+static CHECK_CONSTRAINTS_INFO: FeatureInfo = FeatureInfo {
+    name: "checkConstraints",
+    min_reader_version: 1,
+    min_writer_version: 3,
+    feature_type: FeatureType::Writer,
+    feature_requirements: &[],
+    read_support: KernelSupport::NotSupported,
+    write_support: KernelSupport::NotSupported,
+    enablement_check: EnablementCheck::AlwaysIfSupported,
+};
+
+#[allow(dead_code)]
+static CHANGE_DATA_FEED_INFO: FeatureInfo = FeatureInfo {
+    name: "changeDataFeed",
+    min_reader_version: 1,
+    min_writer_version: 4,
+    feature_type: FeatureType::Writer,
+    feature_requirements: &[],
+    read_support: KernelSupport::NotSupported,
+    write_support: KernelSupport::NotSupported,
```

Comment:
We support reading tables with CDF for read and write. Just that for writes we want appendOnly enabled.


```suggestion
    read_support: KernelSupport::Supported,
    write_support: KernelSupport::Supported,
```